### PR TITLE
[feat]: [PL-25117]: Change the queued on time of the full sync job in rename operation to avoid race conditions

### PR DESCRIPTION
**Pull request by DeepakPatankar**
## Describe your changes

 In rename operation, we are creating two changesets,
      1. The First one to perform rename
      2. Second to perform a full sync
If two renames are done at the same time, then the order of these operations can be
      rename1 rename2 fullsync1 fullsync2
This leads to a bad state in git as fullsync1 will recreate the renamed entity
The solution was to ensure that full sync happens just after rename. So we used the queue on
field to say that full sync is queued just after the rename

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/34818)

### CODE REVIEW

**Review by ershadalim — APPROVED** [review]

File: 400-rest/src/main/java/software/wings/service/impl/yaml/YamlChangeSetHelper.java [hunk 1]
```diff
@@ -138,6 +138,7 @@ private <T> void nonLeafEntityRenameYamlChange(String accountId, T oldEntity, T
     changeSet.addAll(
         entityUpdateService.obtainEntityGitSyncFileChangeSet(accountId, null, newEntity, ChangeType.MODIFY));
     YamlChangeSet savedChangeSet = yamlChangeSetService.saveChangeSet(accountId, changeSet, newEntity);
+    long timeToProcessTheRename = System.currentTimeMillis();
```
Comment (abhinav-harness) [inline 400-rest/src/main/java/software/wings/service/impl/yaml/YamlChangeSetHelper.java:141]: should it timeOfRename 

to process sounds like processing time

Comment (DeepakPatankar) [reply 400-rest/src/main/java/software/wings/service/impl/yaml/YamlChangeSetHelper.java:141]: done

**Review by bhavya181 — APPROVED** [review]
> Reviewed 2 of 3 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34818)* status: 3 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by abhinav-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34818)* status: 3 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/service/impl/yaml/YamlChangeSetHelper.java` line 155 at r3](https://reviewable.io/reviews/harness/harness-core/34818#-N5p4feF8YF-HdulNMSg:-N5p4feG3QS_oKJX8dFE:b-1vzrfq) ([raw file](https://github.com/harness/harness-core/blob/c2bb203e1635084c6d78736080773ae705118ddb/400-rest/src/main/java/software/wings/service/impl/yaml/YamlChangeSetHelper.java#L155)):*
> ```Java
>       // 1. The First one to perform rename
>       // 2. Second to perform a full sync
>       // If two renames are done at the same time, then the order of this operations can be
> ```

multiple doubts: 
1. Are we sure we dont set queued on again at svc layer?
2. Isn't this leaf things inside async block ? - which will again lead to race

**Review by DeepakPatankar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34818)* status: 3 file(s) to review. 1 change requests. 2. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/service/impl/yaml/YamlChangeSetHelper.java` line 155 at r3](https://reviewable.io/reviews/harness/harness-core/34818#-N5p4feF8YF-HdulNMSg:-N5sJRp22VLP_ijNYNPZ:b-xe09al) ([raw file](https://github.com/harness/harness-core/blob/c2bb203e1635084c6d78736080773ae705118ddb/400-rest/src/main/java/software/wings/service/impl/yaml/YamlChangeSetHelper.java#L155)):*
<details><summary><i>Previously, abhinav-harness (Abhinav Singh) wrote…</i></summary><blockquote>

multiple doubts: 
1. Are we sure we dont set queued on again at svc layer?
2. Isn't this leaf things inside async block ? - which will again lead to race
</blockquote></details>

1. I checked that we don't set it again
2. Didn't got the second issue. Please ellaborate

**Review by abhinav-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34818)* status: 3 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/service/impl/yaml/YamlChangeSetHelper.java` line 155 at r3](https://reviewable.io/reviews/harness/harness-core/34818#-N5p4feF8YF-HdulNMSg:-N5sPRqFCy631-u2gsco:bg2hkj7) ([raw file](https://github.com/harness/harness-core/blob/c2bb203e1635084c6d78736080773ae705118ddb/400-rest/src/main/java/software/wings/service/impl/yaml/YamlChangeSetHelper.java#L155)):*
<details><summary><i>Previously, DeepakPatankar (Deepak Patankar) wrote…</i></summary><blockquote>

1. I checked that we don't set it again
2. Didn't got the second issue. Please ellaborate
</blockquote></details>

What I mean is the issue came in since two async threads were competing so we should populate time before async block starts. So is it the case that time is set inside async block?

**Review by DeepakPatankar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34818)* status: 3 file(s) to review. 1 change requests. 2. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/service/impl/yaml/YamlChangeSetHelper.java` line 155 at r3](https://reviewable.io/reviews/harness/harness-core/34818#-N5p4feF8YF-HdulNMSg:-N5siTCICaA9kySSlqfE:b-fwx61e) ([raw file](https://github.com/harness/harness-core/blob/c2bb203e1635084c6d78736080773ae705118ddb/400-rest/src/main/java/software/wings/service/impl/yaml/YamlChangeSetHelper.java#L155)):*
<details><summary><i>Previously, abhinav-harness (Abhinav Singh) wrote…</i></summary><blockquote>

What I mean is the issue came in since two async threads were competing so we should populate time before async block starts. So is it the case that time is set inside async block?
</blockquote></details>

I thought of that solution, one other thing which we have to ensure is that the time of full sync should be more than the time of rename operation.

Hence I am keeping  the time of full sync as the time just after the rename operation

**Review by abhinav-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34818)* status: 3 file(s) to review. 1 change requests. 1 of 2 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/service/impl/yaml/YamlChangeSetHelper.java` line 155 at r3](https://reviewable.io/reviews/harness/harness-core/34818#-N5p4feF8YF-HdulNMSg:-N5tB7op9IygtL9x_y0A:bs7eq6y) ([raw file](https://github.com/harness/harness-core/blob/c2bb203e1635084c6d78736080773ae705118ddb/400-rest/src/main/java/software/wings/service/impl/yaml/YamlChangeSetHelper.java#L155)):*
<details><summary><i>Previously, DeepakPatankar (Deepak Patankar) wrote…</i></summary><blockquote>

I thought of that solution, one other thing which we have to ensure is that the time of full sync should be more than the time of rename operation.

Hence I am keeping  the time of full sync as the time just after the rename operation
</blockquote></details>

if its working after testing i am fine. approving

**Review by abhinav-harness — APPROVED** [review]
> Reviewed 2 of 3 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34818)* status: :shipit: complete! ✓

**Pull request merged**